README.md
# node-port-scanner

[![npm](https://img.shields.io/npm/v/node-port-scanner.svg)](https://www.npmjs.com/package/node-port-scanner)

Scans ports to see if they are open or closed. 

After calling nodePortScanner, ports are checked in order one after the previous is completed.

Multiple calls can be made and will run in parallel.

## Install

```bash
npm i node-port-scanner
```

## Usage

```sh
npm run usage
```

```javascript
const nodePortScanner = require('node-port-scanner');

// scan local common ports
nodePortScanner('127.0.0.1', [21, 22, 23, 25, 80, 110, 123, 443])
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log(error);
  });

// scan remote common ports
nodePortScanner('github.com', [21, 22, 23, 25, 80, 110, 123, 443])
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log(error);
  });

// scan all local ports - not recommended on remote hosts
async function checkLocalPorts () {
  
  const allPorts = nodePortScanner('127.0.0.1', []);
  console.log(await allPorts);
  
}
checkLocalPorts();
```

## Sample Response
```
{
  host: 'github.com',
  ports: { 
    open: [ 22, 80, 443 ], 
    closed: [ 21, 23, 25, 110, 123 ] 
  }
}
```

## Test (Jest)

```sh
npm test
```

## License (MIT)

[MIT](LICENSE)

CHANGELOG.md
# 3.0.0
* [[`7d9b47044c47f32678673ff0fcf095ac565a9097`](https://github.com/beaudurrant/node-port-scanner/commit/7d9b47044c47f32678673ff0fcf095ac565a9097)] - Removed status from the call, added open/closed ports to the response (Beau Durrant)

# 2.2.1
* [[`f67fcb592ef87f44cd3f7f0497bbcfcd41688c81`](https://github.com/beaudurrant/node-port-scanner/commit/f67fcb592ef87f44cd3f7f0497bbcfcd41688c81)] - Updated readme and usage errors (Beau Durrant)

# 2.2.0
* [[`ba3efc022ddfe1a96d42cd8d1b65fb429d4700e8`](https://github.com/beaudurrant/node-port-scanner/commit/ba3efc022ddfe1a96d42cd8d1b65fb429d4700e8)] - Return error objects rather than strings (Beau Durrant)

# 2.1.0
* [[`015fcf731cd046c503f0391d18794971e59e6353`](https://github.com/beaudurrant/node-port-scanner/commit/015fcf731cd046c503f0391d18794971e59e6353)] - Updated tests to use jest (Beau Durrant)

# 2.0.0
* [[`0313d1f87fe9ff2307cead5c3f8f8638556224eb`](https://github.com/beaudurrant/node-port-scanner/commit/0313d1f87fe9ff2307cead5c3f8f8638556224eb)] - Updated to use promise so async/await can be used. (Beau Durrant)

# 1.0.4
* [[`d7b5d427762ed81bcda2bf7f7f3a58a46d730f91`](https://github.com/beaudurrant/node-port-scanner/commit/d7b5d427762ed81bcda2bf7f7f3a58a46d730f91)] - Added circleci config & updated README (Beau Durrant)

# 1.0.3
* [[`45cef9e9293683f2441bc2caaba8c6c6b16fbf94`](https://github.com/beaudurrant/node-port-scanner/commit/45cef9e9293683f2441bc2caaba8c6c6b16fbf94)] - Updated README & version number (Beau Durrant)

# 1.0.2
* [[`a0df2beeb9c6bd4fd6537b18d8f602fb01774dc8`](https://github.com/beaudurrant/node-port-scanner/commit/a0df2beeb9c6bd4fd6537b18d8f602fb01774dc8)] - Updated version number (Beau Durrant)
* [[`920bd383676ab3f31942968c827e004cf3b2ddab`](https://github.com/beaudurrant/node-port-scanner/commit/920bd383676ab3f31942968c827e004cf3b2ddab)] - Added spaces to comments (Beau Durrant)

# 1.0.1
* [[`a0df360e0f4ae4a2672539c388d7388a1965ede5`](https://github.com/beaudurrant/node-port-scanner/commit/a0df360e0f4ae4a2672539c388d7388a1965ede5)] - Updated version number (Beau Durrant)
* [[`e29732a1a050d4b58273d95607d13b7dd5355b7e`](https://github.com/beaudurrant/node-port-scanner/commit/e29732a1a050d4b58273d95607d13b7dd5355b7e)] - Added .gitignore & renamed README (Beau Durrant)

# 1.0.0
* [[`4822b9700e963dbb4bd9a66858de58bab1538a68`](https://github.com/beaudurrant/node-port-scanner/commit/4822b9700e963dbb4bd9a66858de58bab1538a68)] - Initial commit (Beau Durrant)
* [[`e46aaa9c858809a9b0eba727b36e24a526e6f7e5`](https://github.com/beaudurrant/node-port-scanner/commit/e46aaa9c858809a9b0eba727b36e24a526e6f7e5)] - Initial commit (Beau Durrant)

src/tests.js
const nodePortScanner = require('./node-port-scanner.js');

/**
 * ******************************
 *        ERROR CHECKING
 * ******************************
 */

console.log("\n***** ERROR CHECKING *****\n");

nodePortScanner('127.0.0.1', [21, 22, 25, 80, 110, 123, 443], 'not-open-or-closed')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log('Error: ' + error);
  });

nodePortScanner('127.0.0.1', 'not an array!', 'open')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log('Error: ' + error);
  });

nodePortScanner('127.0.0.1', ['a', 'b', 'c'], 'open')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log('Error: ' + error);
  });

nodePortScanner('127.0.0.1', [0], 'open')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log('Error: ' + error);
  });

nodePortScanner('127.0.0.1', [65536], 'open')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log('Error: ' + error);
  });


/**
 * ******************************
 *       CHECK RESPONSES
 * ******************************
 */

console.log("\n***** CHECK RESPONSES *****\n");

// scan for open local common ports
nodePortScanner('127.0.0.1', [21, 22, 23, 25, 80, 110, 123, 443], 'open')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log('Error: ' + error);
  });

// scan for open remote common ports
nodePortScanner('github.com', [21, 22, 23, 25, 80, 110, 123, 443], 'open')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log('Error: ' + error);
  });

//scan for closed local common ports
nodePortScanner('127.0.0.1', [21, 22, 23, 25, 80, 110, 123, 443], 'closed')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log('Error: ' + error);
  });

// scan for closed remote common ports
nodePortScanner('github.com', [21, 22, 23, 25, 80, 110, 123, 443], 'closed')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log('Error: ' + error);
  });


// make calls in parallel
async function checkLocalPorts () {
  
  const openPorts = nodePortScanner('127.0.0.1', [], 'open');
  const closedPorts = nodePortScanner('127.0.0.1', [], 'closed');
  
  console.log(await openPorts);
  console.log(await closedPorts);
  
}
checkLocalPorts();


src/__tests__/node-port-scanner.response.test.js
const nodePortScanner = require('../node-port-scanner.js');

it('check local response', async () => {
  const data = await nodePortScanner('127.0.0.1', []);
  expect(data).toHaveProperty('host', '127.0.0.1');
  expect(data).toHaveProperty('ports');
});

it('check remote response', async () => {
  const data = await nodePortScanner('github.com', [21, 22, 23, 25, 80, 110, 123, 443]);
  expect(data).toHaveProperty('host', 'github.com');
  expect(data).toHaveProperty('ports');
});

src/__tests__/node-port-scanner.validation.test.js
const nodePortScanner = require('../node-port-scanner.js');

it('require host', async () => {
  expect.assertions(1);
  try {
    await nodePortScanner('', [21, 22, 25, 80, 110, 123, 443], 'open');
  } catch (e) {
    expect(e).toEqual({ 'error' : 'host is required' });
  }
});

it('require port status', async () => {
  expect.assertions(1);
  try {
    await nodePortScanner('127.0.0.1', [21, 22, 25, 80, 110, 123, 443], '');
  } catch (e) {
    expect(e).toEqual({ 'error' : 'status is required' });
  }
});

it('validate port status (open/closed)', async () => {
  expect.assertions(1);
  try {
    await nodePortScanner('127.0.0.1', [21, 22, 25, 80, 110, 123, 443], 'not-open-or-closed');
  } catch (e) {
    expect(e).toEqual({ 'error' : 'status must be open or closed' });
  }
});
 
it('require ports', async () => {
  expect.assertions(1);
  try {
    await nodePortScanner('127.0.0.1', '', 'open');
  } catch (e) {
    expect(e).toEqual({ 'error' : 'ports is required' });
  }
});

it('validate ports is an array', async () => {
  expect.assertions(1);
  try {
    await nodePortScanner('127.0.0.1', 'not an array!', 'open');
  } catch (e) {
    expect(e).toEqual({ 'error' : 'ports must be an array' });
  }
});

it('validate ports are integers', async () => {
  expect.assertions(1);
  try {
    await nodePortScanner('127.0.0.1', ['a', 'b', 'c'], 'open');
  } catch (e) {
    expect(e).toEqual({ 'error' : 'port must be an integer' });
  }
});

it('validate ports in range', async () => {
  expect.assertions(1);
  try {
    await nodePortScanner('127.0.0.1', [0], 'open');
  } catch (e) {
    expect(e).toEqual({ 'error' : 'port must be in range [1-65535]' });
  }
});

it('validate ports in range', async () => {
  expect.assertions(1);
  try {
    await nodePortScanner('127.0.0.1', [65536], 'open');
  } catch (e) {
    expect(e).toEqual({ 'error' : 'port must be in range [1-65535]' });
  }
});

src/node-port-scanner.js
// what we need to use sockets
const net = require('net');

// scan range of ports for status (open|closed)
module.exports = nodePortScanner = (host, ports, status) => {

  return new Promise((resolve, reject) => {
    
    // all ports that exist
    const allPorts = Array.from({length: 65535}, (_, i) => i + 1);
  
    // connect to a single port and get the status
    const connectToPort = (host, port, callback) => {
      
      // error checking args
      if (!Number.isInteger(port)) reject({ 'error' : 'port must be an integer' });
      if (port < 1 || port > 65535) reject({ 'error' : 'port must be in range [1-65535]' });
      
      let socket = new net.Socket();
      // increase this if y'all on dial up
      let timeout = 200;
      
      // new properties & events for port scanner
      socket._scan = {};
      socket._scan.status = 'initialized';
      socket._scan.host = host;
      socket._scan.port = port;
      socket._scan._events = { complete : callback };
      
      // events for socket
      socket.on('connect', function () {
        this._scan.status = 'connect';
        socket.destroy();
      });
      socket.on('timeout', function () {
        this._scan.status = 'timeout';
        socket.destroy();
      });
      socket.on('error', function (exception) {
        this._scan.status = 'error';
        socket.destroy();
      });
      socket.on('close', function (exception) {
        this._scan._events.complete(this._scan);
      });
      
      socket.setTimeout(timeout);
      socket.connect(port, host);
      
    };
  
    // recursive function to check all port status one after the other is complete
    const connectToPorts = (host, ports, status, scanResults) => {
      
      let port = ports.shift();
      
      connectToPort(host, port, function (result) {
        
        // add to our results based on the status of the result and scan
        if ((result.status == 'connect' && status == 'open') ||
            (result.status != 'connect' && status == 'closed')) {
          scanResults.ports.push(result.port);
        }
        
        // recursivly go through all the ports
        if (ports.length) {
          connectToPorts(host, ports, status, scanResults);
        }
        // when ports are done resolve the promise
        else {
          resolve(scanResults);
        }
        
      });
      
    };
    
    // error checking args
    if (host == undefined || !host) reject({ 'error' : 'host is required' });
    if (ports == undefined || !ports) reject({ 'error' : 'ports is required' });
    if (!Array.isArray(ports)) reject({ 'error' : 'ports must be an array' });
    if (status == undefined || !status) reject({ 'error' : 'status is required' });
    if (status.toLowerCase() != 'open' && status.toLowerCase() != 'closed') reject({ 'error' : 'status must be open or closed' });

    // scan results will be and array of port numbers matching status
    let scanResults = { host : host, status : status, ports : [] };
  
    // no ports = all ports
    if (!ports.length) ports = allPorts;
    
    connectToPorts(host, ports, status, scanResults);
  
  });
  
};


src/usage.js
const nodePortScanner = require('./node-port-scanner.js');

// scan for open local common ports
nodePortScanner('127.0.0.1', [21, 22, 23, 25, 80, 110, 123, 443], 'open')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log(error);
  });

// scan for open remote common ports
nodePortScanner('github.com', [21, 22, 23, 25, 80, 110, 123, 443], 'open')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log(error);
  });

// scan for closed local common ports
nodePortScanner('127.0.0.1', [21, 22, 23, 25, 80, 110, 123, 443], 'closed')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log(error);
  });

// scan for closed remote common ports
nodePortScanner('github.com', [21, 22, 23, 25, 80, 110, 123, 443], 'closed')
  .then(results => {  
    console.log(results);
  })
  .catch(error => {
    console.log(error);
  });


// make calls in parallel
async function checkLocalPorts () {
  
  const openPorts = nodePortScanner('127.0.0.1', [], 'open');
  const closedPorts = nodePortScanner('127.0.0.1', [], 'closed');
  
  console.log(await openPorts);
  console.log(await closedPorts);
  
}
checkLocalPorts();
